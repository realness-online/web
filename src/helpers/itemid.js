// {base_url}/{:author}/{:type}/{:created_at}
import firebase from 'firebase/app'
import 'firebase/storage'
import 'firebase/auth'
import get_item from '@/modules/item'
import { does_not_exist } from '@/persistance/Cloud.sync'
import { get, set } from 'idb-keyval'
// Expensive to call
export async function load(itemid, me = localStorage.me) {
  let item
  if (~itemid.indexOf(me)) {
    item = localStorage.getItem(itemid)
    if (item) return get_item(item)
    else if (as_type(itemid) === 'relations') return []
  }
  const result = await get(itemid)
  item = get_item(result)
  if (item) return item
  item = await load_from_network(itemid, me)
  if (item) return item
  return null
}
export async function list(itemid, me = localStorage.me) {
  try {
    const item = await load(itemid, me)
    if (item) return type_as_list(item)
    else return []
  } catch {
    return []
  }
}
export async function load_from_network(itemid, me = localStorage.me) {
  const url = await as_download_url(itemid, me)
  if (url) {
    console.info('download', itemid)
    const server_text = await (await fetch(url)).text()
    await set(itemid, server_text)
    return get_item(server_text)
  } else return null
}
export async function as_directory(itemid) {
  const path = as_directory_id(itemid)
  const cached = await get(path)
  if (cached) return cached
  else if (navigator.onLine && firebase.auth().currentUser) {
    const meta = { items: [], types: [] } // folders are types in our vocabulary
    console.info('request:directory', itemid)
    const directory = await firebase.storage().ref().child(`people/${path}`).listAll()
    directory.items.forEach(item => meta.items.push(item.name.split('.')[0]))
    directory.prefixes.forEach(prefix => meta.types.push(prefix.name))
    await set(path, meta)
    return meta
  } else return null
}
export async function as_download_url(itemid) {
  if (!firebase.auth().currentUser) return null
  const storage = firebase.storage().ref()
  try {
    console.info('request:location', itemid)
    return await storage.child(as_filename(itemid)).getDownloadURL()
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      console.warn(itemid, '=>', as_filename(itemid))
      const index = (await get('sync:index')) || {}
      index[itemid] = does_not_exist
      await set('sync:index', index)
      return null
    } else throw e
  }
}

// Cheap to call
const created_at = ['avatars', 'posters', 'animations']
const has_history = ['statements', 'events']
export function is_history(itemid) {
  const parts = as_path_parts(itemid)
  if (has_history.includes(as_type(itemid)) && parts.length === 3) return true
  return false
}
export function as_filename(itemid) {
  let filename = itemid
  if (itemid.startsWith('/+')) filename = `/people${filename}`
  if (created_at.includes(as_type(itemid)) || is_history(itemid)) return `${filename}.html`
  else return `${filename}/index.html`
}
export function as_storage_path(itemid) {
  const path_parts = as_path_parts(itemid)
  let path = `/${path_parts[0]}`
  if (itemid.startsWith('/+')) path = `/people${path}`
  switch (path_parts.length) {
    case 0:
      return null
    case 1:
      return path
    default:
      return `${path}/${path_parts[1]}`
  }
}
export function as_directory_id(itemid) {
  const parts = as_path_parts(itemid)
  return `/${parts[0]}/${parts[1]}/`
}
export function as_path_parts(itemid) {
  const path = itemid.split('/')
  if (path[0].length === 0) path.shift()
  return path
}
export function as_author(itemid) {
  const path = as_path_parts(itemid)
  const author = path[0] || ''
  if (author.startsWith('+1')) return `/${path[0]}`
  else return null
}
export function as_type(itemid) {
  const path = as_path_parts(itemid)
  if (path[1]) return path[1]
  if (itemid.startsWith('/+')) return 'person'
  else return null
}
export function as_created_at(itemid) {
  const path = as_path_parts(itemid)
  return parseInt(path[2])
}
export function as_query_id(itemid) {
  return itemid.substring(2).replace('/', '-').replace('/', '-')
}
export function as_fragment(itemid) {
  return `#${as_query_id(itemid)}`
}
export function type_as_list(item) {
  // Returns a list even if loading the item fails
  // the microdata spec requires properties values to
  // single value and iterable
  if (!item) return []
  const list = item[as_type(item.id)]
  if (list && Array.isArray(list)) return list
  else if (list) return [list]
  else return []
}
