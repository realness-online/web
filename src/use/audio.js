import { ref } from 'vue'

/**
 * @typedef {Object} AudioAnalyzer
 * @property {boolean} analyzing - Whether audio analysis is currently active
 * @property {(stream: MediaStream) => void} analyze_audio - Start analyzing audio from stream
 */

/**
 * Audio analysis composable
 * @returns {AudioAnalyzer}
 */
export const use = () => {
  /** @type {import('vue').Ref<AnalyserNode|null>} */
  const analyser = ref(null)

  /** @type {import('vue').Ref<boolean>} */
  const analyzing = ref(false)

  /**
   * Processes audio data on each animation frame
   */
  const process_audio = () => {
    const times = new Uint8Array(analyser.value.frequencyBinCount)
    analyser.value?.getByteTimeDomainData(times)
    // for (var i = 0; i < times.length; i++) {
    //   const value = times[i]
    //   const zerod = value - 128
    //   console.log(i, value, zerod)
    // }
    if (analyzing.value) requestAnimationFrame(process_audio)
  }

  /**
   * Initializes audio analysis for a media stream
   * @param {MediaStream} stream - The audio stream to analyze
   */
  const analyze_audio = stream => {
    const context = new AudioContext()
    analyser.value = context.createAnalyser()
    analyser.value.fftSize = 32
    analyser.value.minDecibels = -64
    analyser.value.maxDecibels = -6
    const input = context.createMediaStreamSource(stream)
    input.connect(analyser.value)
    analyzing.value = true
    requestAnimationFrame(process_audio)
  }

  return {
    analyzing,
    analyze_audio
  }
}
