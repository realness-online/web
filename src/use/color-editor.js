export function change_color(id, color, type = 'fill') {
  const path = this.get_path_element(id)
  path.setAttribute(type, color)
}

export function get_color(id, type = 'fill') {
  return this.get_path_element(id).getAttribute(`${type}`)
}

export function get_path_element(id) {
  return document.getElementById(id).querySelector('path')
}

export function get_path_attribute(id, type = 'fill') {
  return this.get_path_element(id).getAttribute(`${type}`)
}
