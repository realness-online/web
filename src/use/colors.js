import rgb_hex from 'rgb-hex'

// https://una.im/css-color-theming
// takes hex rgb or hsl value and returns hsla value
export function to_blended_hsla(color = '') {
  let H = color.toString()
  // check if it's already hsl
  if (H.startsWith('hsl')) return H
  if (H.startsWith('rgb')) H = rgb_hex(H)
  // Convert hex to RGB first
  let r = 0
  let g = 0
  let b = 0
  if (H.length == 4) {
    r = '0x' + H[1] + H[1]
    g = '0x' + H[2] + H[2]
    b = '0x' + H[3] + H[3]
  } else if (H.length == 7) {
    r = '0x' + H[1] + H[2]
    g = '0x' + H[3] + H[4]
    b = '0x' + H[5] + H[6]
  }
  // Then to HSL
  r /= 255
  g /= 255
  b /= 255
  let cmin = Math.min(r, g, b)
  let cmax = Math.max(r, g, b)
  let delta = cmax - cmin
  let h = 0
  let s = 0
  let l = 0

  if (delta == 0) h = 0
  else if (cmax == r) h = ((g - b) / delta) % 6
  else if (cmax == g) h = (b - r) / delta + 2
  else h = (r - g) / delta + 4

  h = Math.round(h * 60)

  if (h < 0) h += 360

  l = (cmax + cmin) / 2
  s = delta == 0 ? 0 : delta / (1 - Math.abs(2 * l - 1))
  s = +(s * 100).toFixed(1)
  l = +(l * 100).toFixed(1)

  return `hsla(${h},${s}%,${l}%,0.95)`
}
