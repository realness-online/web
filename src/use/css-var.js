export default function(name) {
  return getComputedStyle(document.documentElement).getPropertyValue(name)
}
