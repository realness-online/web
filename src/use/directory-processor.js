import { use as use_vectorize } from '@/use/vectorize'
import { use as use_optimizer } from '@/use/optimizer'
import { ref } from 'vue'

const use_directory_processor = () => {
  const { new_vector, new_gradients, process_photo } = use_vectorize()
  const { optimize } = use_optimizer(completed_poster)
  const progress = ref({
    total: 0,
    current: 0,
    processing: false,
    current_file: ''
  })
  const current_preview = ref(null)
  const completed_poster = ref(null)

  const process_directory = async () => {
    try {
      progress.value.processing = true
      current_preview.value = null
      completed_poster.value = null

      console.info('üóÇÔ∏è Starting directory processing...')

      const source_dir = await window.showDirectoryPicker({
        mode: 'read',
        startIn: 'pictures',
        id: 'source-images'
      })

      // Count total image files
      let image_count = 0
      for await (const [name, handle] of source_dir.entries()) {
        if (handle.kind === 'file' && name.match(/\.(jpg|jpeg|png|gif|webp)$/i)) {
          image_count++
        }
      }
      progress.value.total = image_count
      console.info(`üìÇ Found ${image_count} images to process`)

      const posters_dir = await source_dir.getDirectoryHandle('posters', { create: true })

      for await (const [name, handle] of source_dir.entries()) {
        if (handle.kind !== 'file') continue
        if (!name.match(/\.(jpg|jpeg|png|gif|webp)$/i)) continue

        progress.value.current_file = name
        console.info(`üñºÔ∏è Processing file: ${name}`)

        const file = await handle.getFile()
        const image_url = URL.createObjectURL(file)

        try {
          await process_photo(image_url)
          while (!new_vector.value) {
            await new Promise(r => setTimeout(r, 100))
          }

          completed_poster.value = new_vector.value

          if (!completed_poster.value.optimized) {
            await optimize()
          }

          const svg_data = completed_poster.value
          const poster_name = name.replace(/\.[^/.]+$/, '.svg')
          const poster_file = await posters_dir.getFileHandle(poster_name, { create: true })

          const writable = await poster_file.createWritable()
          await writable.write(svg_data)
          await writable.close()

          new_vector.value = null
          if (new_gradients.value) new_gradients.value = null

          progress.value.current++
          console.info(`‚úÖ Saved poster: ${poster_name}`)
        } catch (error) {
          console.error(`‚ùå Failed to process ${name}:`, error)
        } finally {
          URL.revokeObjectURL(image_url)
        }
      }

      console.info('üéâ Directory processing complete!')
    } catch (error) {
      console.error('‚ùå Directory processing failed:', error)
    } finally {
      progress.value.processing = false
      progress.value.current = 0
      progress.value.total = 0
      progress.value.current_file = ''
      completed_poster.value = null
    }
  }

  return {
    process_directory,
    progress,
    completed_poster
  }
}

export { use_directory_processor }
