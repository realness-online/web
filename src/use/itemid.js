// {base_url}/{:author}/{:type}/{:created_at}
import get_item from '@/use/item'
import { does_not_exist } from '@/use/sync'
import { current_user, url, directory } from '@/use/serverless'
import { get, set, keys } from 'idb-keyval'
class Directory {
  constructor() {
    this.items = []
    this.types = [] // folders are types in our vocabulary
  }
}
// Expensive to call
export async function load(itemid, me = localStorage.me) {
  let item
  if (~itemid.indexOf(me)) {
    item = localStorage.getItem(itemid)
    if (item) return get_item(item)
    else if (as_type(itemid) === 'relations') return []
  }
  const result = await get(itemid)
  item = get_item(result)
  if (item) return item
  item = await load_from_network(itemid, me)
  if (item) return item
  return null
}
export async function list(itemid, me = localStorage.me) {
  try {
    const item = await load(itemid, me)
    if (item) return type_as_list(item)
    else return []
  } catch {
    return []
  }
}
export async function load_from_network(itemid, me = localStorage.me) {
  const url = await as_download_url(itemid, me)
  if (url) {
    console.info('download', itemid)
    const server_text = await (await fetch(url)).text()
    await set(itemid, server_text)
    return get_item(server_text)
  } else return null
}
export async function load_directory_from_network(itemid) {
  if (navigator.onLine && current_user.value) {
    const path = as_directory_id(itemid)
    const meta = new Directory()
    console.info('request:directory', itemid)
    const folder = await directory(`people/${path}`)
    folder.items.forEach(item => meta.items.push(item.name.split('.')[0]))
    folder.prefixes.forEach(prefix => meta.types.push(prefix.name))
    await set(path, meta)
    return meta
  }
}
export async function as_directory(itemid) {
  const path = as_directory_id(itemid)
  const cached = await get(path)
  if (cached) return cached
  let directory = await build_local_directory(itemid)
  if (navigator.onLine && current_user.value)
    directory = await load_directory_from_network(itemid)
  return directory
}
export async function as_download_url(itemid) {
  if (!current_user.value) return null
  try {
    console.info('request:location', itemid)
    return await url(as_filename(itemid))
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      console.warn(itemid, '=>', as_filename(itemid))
      const index = (await get('sync:index')) || {}
      index[itemid] = does_not_exist
      await set('sync:index', index)
      return null
    } else throw e
  }
}

// Cheap to call
const created_at = ['posters', 'animations']
const has_history = ['statements', 'events']
export function is_history(itemid) {
  const parts = as_path_parts(itemid)
  if (has_history.includes(as_type(itemid)) && parts.length === 3) return true
  else return false
}
export function as_filename(itemid) {
  let filename = itemid
  if (itemid.startsWith('/+')) filename = `/people${filename}`
  if (created_at.includes(as_type(itemid)) || is_history(itemid))
    return `${filename}.html`
  else return `${filename}/index.html`
}
export function as_storage_path(itemid) {
  const path_parts = as_path_parts(itemid)
  let path = `/${path_parts[0]}`
  if (itemid.startsWith('/+')) path = `/people${path}`
  switch (path_parts.length) {
    case 0:
      return null
    case 1:
      return path
    default:
      return `${path}/${path_parts[1]}`
  }
}
export async function build_local_directory(itemid) {
  const path = as_directory_id(itemid)
  const directory = new Directory()
  const everything = await keys()
  everything.forEach(itemid => {
    if (as_directory_id(itemid) === path) {
      const id = as_created_at(itemid)
      if (id) directory.items.push(id)
    }
  })
  return directory
}
export function as_directory_id(itemid) {
  const parts = as_path_parts(itemid)
  return `/${parts[0]}/${parts[1]}/`
}
export function as_path_parts(itemid) {
  const path = itemid.split('/')
  if (path[0].length === 0) path.shift()
  return path
}
export function as_author(itemid) {
  const path = as_path_parts(itemid)
  const author = path[0] || ''
  if (author.startsWith('+1')) return `/${path[0]}`
  else return null
}
export function as_type(itemid) {
  const path = as_path_parts(itemid)
  if (path[1]) return path[1]
  if (itemid.startsWith('/+')) return 'person'
  else return null
}
export function as_created_at(itemid) {
  const path = as_path_parts(itemid)
  if (path.length === 1) return parseInt(path[0])
  return parseInt(path[2])
}
export function as_query_id(itemid) {
  return itemid.substring(2).replace('/', '-').replace('/', '-')
}
export function as_fragment_id(itemid) {
  return `#${as_query_id(itemid)}`
}
export function type_as_list(item) {
  // Returns a list even if loading the item fails
  // the microdata spec requires properties values to
  // single value and iterable
  if (!item) return []
  const list = item[as_type(item.id)]
  if (list && Array.isArray(list)) return list
  else if (list) return [list]
  else return []
}
