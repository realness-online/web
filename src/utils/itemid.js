/** @typedef {import('@/types').Id} Id */
/** @typedef {import('@/types').Type} Type */
/** @typedef {import('@/types').Created} Created */
/** @typedef {import('@/types').Author} Author */
/** @typedef {import('@/types').Item} Item */
import { get, set } from 'idb-keyval'
import get_item from '@/utils/item'
import { DOES_NOT_EXIST } from '@/use/sync'
import { url } from '@/utils/serverless'
import { decompress_html } from '@/utils/upload-processor'
import { as_directory_id } from '@/persistance/Directory'
import { newest_timestamp_first } from '@/utils/sorting'
/**
 * @type {Type[]}
 */
const has_archive = ['posters']

/**
 * @type {Type[]}
 */
const has_history = ['statements', 'events']

// this leaves me, history, relations, and offline

/**
 * @param {Id} itemid
 * @returns {Promise<string>}
 */
export const as_filename = async itemid => {
  let filename = itemid
  if (itemid.startsWith('/+')) filename = `people${itemid}`

  if (has_archive.includes(as_type(itemid))) {
    const archive = await as_archive(itemid)
    console.log('archive name', archive)
    if (archive) return `${archive}.html.gz`
    return `${filename}.html.gz`
  } else if (is_history(itemid)) return `${filename}.html.gz`

  return `${filename}/index.html.gz`
}

/**
 * @param {Id} itemid
 * @returns {Promise<string | null>}
 */
export const load_from_network = async itemid => {
  const url = await as_download_url(itemid)

  if (url) {
    const response = await fetch(url)

    // Check Content-Encoding header
    const content_encoding = response.headers.get('Content-Encoding')
    const compressed_html = await response.arrayBuffer()
    let html = null
    // If no content encoding or 'identity', data is already decompressed
    if (!content_encoding || content_encoding === 'identity')
      html = new TextDecoder().decode(compressed_html)
    else html = await decompress_html(compressed_html)

    if (!html) return null
    console.log('storage', itemid)
    await set(itemid, html)
    return get_item(html)
  }
  return null
}

/**
 * @param {Id} itemid
 * @param {Author} me
 * @returns {Promise<Item | null>}
 */
export const load = async (itemid, me = localStorage.me) => {
  console.trace('load', itemid)
  let item
  if (~itemid.indexOf(me)) {
    item = localStorage.getItem(itemid)
    if (item) return get_item(item)
    else if (as_type(itemid) === 'relations') return []
  }
  const result = await get(itemid)
  item = get_item(result)
  if (item) return item
  try {
    item = await load_from_network(itemid, me)
  } catch (e) {
    if (e.code === 'storage/unauthorized') return null
    throw e
  }
  if (item) return item
  return null
}

/**
 * @param {Id} itemid
 * @param {Author} me
 * @returns {Promise<Type[]>}
 */
export const list = async (itemid, me = localStorage.me) => {
  try {
    const item = await load(itemid, me)
    if (item) return type_as_list(item)
    return []
  } catch {
    return []
  }
}

/**
 * @param {Id} itemid
 * @returns {Promise<string | null>}
 */
export const as_download_url = async itemid => {
  if (itemid.startsWith('/+/')) return null
  try {
    return await url(await as_filename(itemid))
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      console.trace(itemid, '=>', await as_filename(itemid))
      const index = (await get('sync:index')) || {}
      index[itemid] = DOES_NOT_EXIST
      await set('sync:index', index)
      return null
    }
    throw e
  }
}

/**
 * Splits an item ID path into its constituent parts
 * @param {Id} itemid - The full item path (e.g. '/+123456/statements/789')
 * @returns {[Author] | [Author, Type] | [Author, Type, Created]} Array of path parts where:
 *   [0] = author ID (e.g. '+123456')
 *   [1] = item type (e.g. 'statements', 'events', 'posters', etc.)
 *   [2] = created_at timestamp (if applicable)
 * @example
 * as_path_parts('/+123456/statements/789') // ['+123456', 'statements', '789']
 * as_path_parts('/+123456') // ['+123456']
 * as_path_parts('/+123456/statements/789/') // ['+123456', 'statements', null, '789']
 */
export const as_path_parts = itemid => {
  const path = itemid.split('/')
  if (path[0].length === 0) path.shift()
  if (itemid.endsWith('/')) {
    //is a directory
    const [author, type, created = null, archive = null] = path
    if (!created && !archive) return [author, type, 'index', null]
    if (created) return [author, type, null, created]
  }
  return path
}

/**
 * @param {Id} itemid
 * @returns {string | null}
 */
export const as_storage_path = itemid => {
  const path_parts = as_path_parts(itemid)
  let path = `/${path_parts[0]}`
  if (itemid.startsWith('/+')) path = `/people${path}`
  switch (path_parts.length) {
    case 0:
      return null
    case 1:
      return path
    default:
      return `${path}/${path_parts[1]}`
  }
}

/**
 * @param {Id} itemid
 * @returns {string | null}
 */
export const as_author = itemid => {
  const path = as_path_parts(itemid)
  const author = path[0] || ''
  if (author.startsWith('+1')) return `/${path[0]}`
  return null
}

/**
 * @param {Id} itemid
 * @returns {Type | null}
 */
export const as_type = itemid => {
  const path = as_path_parts(itemid)
  if (path[1]) return path[1]
  if (itemid.startsWith('/+')) return 'person'
  return null
}

/**
 * @param {Id} itemid
 * @returns {Created | null}
 */
export const as_created_at = itemid => {
  const path = as_path_parts(itemid)
  /** @type {1 | 2 | 3 | 4} */
  const path_length = path.length
  if (path_length === 3) return parseInt(path[2])
  if (path_length === 4) return parseInt(path[3])
  return null
}

/**
 * @param {Id} itemid
 * @returns {string}
 */
export const as_query_id = itemid =>
  itemid.substring(2).replace('/', '-').replace('/', '-')
//

/**
 * @param {Id} itemid
 * @returns {string}
 */
export const as_fragment_id = itemid => `#${as_query_id(itemid)}`

/**
 * @param {Item} item
 * @returns {Type[]}
 */
export const type_as_list = item => {
  // Returns a list even if loading the item fails
  // the microdata spec requires properties values to
  // single value and iterable
  if (!item) return []
  const list = item[as_type(item.id)]
  if (list && Array.isArray(list)) return list
  else if (list) return [list]
  return []
}
/**
 * @param {Id} itemid
 * @returns {boolean}
 */
export const is_history = itemid => {
  const parts = as_path_parts(itemid)
  if (has_history.includes(as_type(itemid)) && parts.length === 3) return true
  return false
}

/**
 * Determines if an item needs to be loaded from an archive
 * @param {Id} itemid - The ID of the item to check
 * @returns {Promise<string | null>} - Returns null if the item exists in the main items list,
 *                                    otherwise returns the archive path if found in an archive,
 *                                    or null if not found in either location
 */
export const as_archive = async itemid => {
  const directory = await get(as_directory_id(itemid))
  if (!directory) return null

  const { items = [], archive = [] } = directory
  const created = as_created_at(itemid)
  if (!created) return null

  if (items.includes(created.toString())) return null

  if (archive.includes(created))
    return `people${as_author(itemid)}/${as_type(itemid)}/${created}/${created}`

  const sorted_archive = [...archive].sort(newest_timestamp_first)

  let closest_timestamp = null // Find the closest archive timestamp that's GREATER than or EQUAL to the created timestamp
  for (const timestamp of sorted_archive)
    if (created <= timestamp) {
      closest_timestamp = timestamp
      break // Stop at the first timestamp that's greater than or equal to created
    }

  if (!closest_timestamp) return null
  return `people${as_author(itemid)}/${as_type(itemid)}/${closest_timestamp}/${created}`
}
