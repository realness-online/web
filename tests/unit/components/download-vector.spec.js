import { shallowMount, flushPromises } from '@vue/test-utils'
import download_vector from '@/components/download-vector'
import * as itemid from '@/use/itemid'
import get_item, { hydrate } from '@/use/item'
const fs = require('fs')
const poster_html = fs.readFileSync('./tests/unit/html/poster.html', 'utf8')
const poster = get_item(poster_html)
const person = {
  id: '/+14151234356',
  first_name: 'Scott',
  last_name: 'Fryxell',
  mobile: '4151234356',
  avatar: 'avatars/5553338945763'
}
describe('@/components/download-vector', () => {
  describe('Renders', () => {
    it('Link to download svg', () => {
      jest
        .spyOn(document, 'getElementById')
        .mockImplementation(() => hydrate(poster_html))

      jest
        .spyOn(itemid, 'load')
        .mockImplementation(() => Promise.resolve(person))

      const wrapper = shallowMount(download_vector, {
        props: {
          itemid: poster.id
        }
      })

      expect(wrapper.element).toMatchSnapshot()
    })
    it('Handles downloads from anonymous users', async () => {
      jest
        .spyOn(document, 'getElementById')
        .mockImplementation(() => hydrate(poster_html))

      jest.spyOn(itemid, 'load').mockImplementation(() => Promise.resolve(null))

      const wrapper = shallowMount(download_vector, {
        props: {
          itemid: poster.id
        }
      })
      await flushPromises()
      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
