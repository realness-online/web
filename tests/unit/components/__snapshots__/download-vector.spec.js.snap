// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/components/download-vector Renders link to download svg 1`] = `
<a
  class="download"
  download="Scott_Fryxell_posters_Saturday, September 26, 8:02 AM.svg"
  href="data:application/octet-stream,%3Csvg%20viewBox%3D%220%200%20333%20444%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%3Cpath%20itemprop%3D%22path%22%20d%3D%22M0%20222.003v222.003l166.75-.253%20166.75-.253.252-221.75L334.004%200H0v222.003m.491.497c0%20122.1.117%20171.901.259%20110.67.142-61.232.142-161.132%200-222C.608%2050.301.491%20100.4.491%20222.5M207.5%2082.642c-9.571%201.026-26.64%204.489-30.694%206.227-4.976%202.133-10.466%207.178-12.019%2011.045-1.915%204.769-3.636%2014.498-2.839%2016.05.434.845%201.145%203.336%201.581%205.536.436%202.2.932%204.45%201.103%205%20.172.55.437%201.9.59%203%20.153%201.1.547%202.225.876%202.5.329.275.906%202.469%201.282%204.875.655%204.191.588%204.351-1.598%203.8-2.199-.555-2.212-.523-.341.875%201.388%201.037%201.531%201.454.5%201.465-1.171.013-1.191.174-.107.861%201.064.675.997.975-.332%201.485-2.295.881-2.722.804-2.002-.361.691-1.118.593-1.148-5-1.496-2.2-.136-4.9-.562-6-.945-1.322-.461-1.152-.177.5.836%201.375.844%203.989%201.55%205.809%201.57%201.82.019%203.057.443%202.749.941-.308.498-3.427.836-6.932.75-5.271-.129-6.283.103-5.852%201.344.395%201.139.225%201.229-.705.372-.673-.62-1.599-2.79-2.058-4.823-.672-2.98-.563-3.591.563-3.159.783.301%201.884-.378%202.511-1.548.798-1.493.803-1.891.015-1.404-.605.374-1.169.091-1.252-.629-.084-.72-.218-1.984-.297-2.809-.079-.825-.604-1.415-1.165-1.311-.656.122-.866-2.774-.586-8.093.433-8.254.421-8.302-3.355-13.69-4.42-6.306-5.282-7.072-11.945-10.607-9.932-5.271-35.146-4.003-52.707%202.651-22.902%208.676-53.457%2042.857-62.867%2070.328-9.512%2027.765-4.831%2069.247%2010.311%2091.393%202.697%203.944%209.094%2011.225%2014.216%2016.18%2010.194%209.861%2018.363%2014.678%2033.047%2019.485C85.649%20304.64%20133.4%20314%20142.209%20314c3.035%200%207.904.452%2010.82%201.005%204.535.861%205.577.757%207.207-.718%201.157-1.048%203.153-1.617%205.085-1.452%201.748.151%204.799-.404%206.779-1.231%201.979-.828%2022.904-6.424%2046.5-12.436%2046.73-11.906%2048.1-12.423%2061.626-23.264%2011.305-9.061%2033.207-35.488%2039.007-47.066%205.596-11.172%207.206-18.421%207.069-31.838-.312-30.791-16.918-62.476-45.077-86.014-17.13-14.32-31.345-22.564-44.867-26.022-7.608-1.946-21.753-3.084-28.858-2.322m-5.5%206.944c-24.146%203.186-31.562%207.043-33.479%2017.414-.48%202.592.281%206.375%202.931%2014.581%201.968%206.094%203.931%2013.744%204.363%2017%20.577%204.35%201.169%205.849%202.235%205.655%201.136-.208%201.1.011-.163%201.01-1.427%201.127-1.167%201.556%202.25%203.715%203.469%202.192%205.089%204.233%203.113%203.921-2.06-.326-2.461.246-.96%201.368%201.989%201.487.717%204.35-1.547%203.482-1.405-.54-3.198-5.907-2.833-8.482.098-.687-.244-1.25-.759-1.25-1.322%200-2.405%203.22-1.32%203.92.495.319%201.233%201.649%201.641%202.955.93%202.976-.645%204.089-5.818%204.11-2.115.008-6.351.743-9.412%201.633-4.256%201.237-6.354%201.376-8.917.59-2.892-.888-3.49-.767-4.373.882-.771%201.44-2.06%201.891-5.237%201.831-2.318-.043-3.623-.316-2.9-.605.723-.289%201.73-1.62%202.239-2.958.729-1.917.408-3.22-1.511-6.145-1.34-2.042-2.744-6.076-3.119-8.963-.375-2.887-1.186-5.257-1.803-5.265-.788-.011-.775-.236.046-.758.688-.438%201.46-5.122%201.882-11.421.688-10.267.618-10.846-1.819-15.003-1.532-2.615-4.253-5.304-6.882-6.803-3.829-2.183-5.789-2.521-16.418-2.837-16.212-.482-26.166%201.417-37.93%207.237-10.957%205.42-21.573%2014.301-32.122%2026.872C22.572%20162.067%2015.849%20178.608%2015.851%20205c.001%206.875.526%2014.525%201.167%2017%20.642%202.475%201.268%205.423%201.392%206.55.124%201.128.559%202.591.967%203.25.407.66.472%201.635.143%202.167-.851%201.377%203.49%2015.145%204.566%2014.48.49-.303.614.17.276%201.052-.365.95.293%202.57%201.616%203.978%201.365%201.453%201.934%202.889%201.466%203.699-.42.728-.415%201.009.013.623.428-.385%202.677%201.865%204.998%205%206.105%208.246%2018.316%2019.421%2025.677%2023.499%2012.581%206.969%2035.091%2013.293%2038.883%2010.925%201.471-.919%202.807-.836%206.181.386%2014.275%205.168%2028.562%207.636%2043.433%207.503%209.491-.085%2012.392.208%2013.336%201.346%201.035%201.247%201.515%201.117%203.371-.912%201.733-1.894%205.653-3.173%2019.664-6.415%2029.976-6.935%2035.901-8.106%2037.641-7.438%201.225.47%201.464.287.874-.669-.629-1.018-.195-1.161%201.92-.63%201.847.464%202.516.336%202.065-.394-.451-.73.218-.858%202.065-.394%201.64.411%202.499.306%202.147-.263-.326-.528%205.482-2.459%2013.101-4.354A7138.686%207138.686%200%2000257%20281.451c.275-.072%202.3-.964%204.5-1.981%202.2-1.017%204.33-1.876%204.733-1.909%201.443-.12%2011.861-8.775%2013.414-11.144.87-1.33%202.319-2.424%203.218-2.432%201.299-.012%201.356-.192.278-.875-.946-.6-1.021-1.067-.25-1.544.609-.376%201.557-.406%202.107-.066.55.34%201%20.182%201-.351%200-.533%202.527-3.597%205.617-6.809%2012.526-13.025%2022.101-27.989%2025.81-40.34%202.3-7.659%203.141-23.38%201.667-31.169-3.546-18.734-12.985-37.491-26.474-52.608-6.721-7.532-17.522-17.383-23.569-21.496-2.503-1.702-4.776-3.354-5.051-3.67-2.955-3.401-21.69-12.6-29.5-14.485-8.51-2.053-21.435-2.445-32.5-.986%22%20fill-rule%3D%22evenodd%22%3E%3C%2Fpath%3E%3C%2Fsvg%3E"
>
  <icon-stub
    name="download"
  />
</a>
`;
